src/runtime_action.rs

```diff
@@ -0,0 +1,126 @@
+use crate::auth_action::AuthAction;
+use crate::configuration::{Action, FailureMode, Service, ServiceType};
+use crate::ratelimit_action::RateLimitAction;
+use crate::service::GrpcService;
+use std::collections::HashMap;
+use std::rc::Rc;
+use std::time::Duration;
+
+#[derive(Clone, Debug)]
+pub enum RuntimeAction {
+    Auth(AuthAction),
+    RateLimit(RateLimitAction),
+}
+
+impl RuntimeAction {
+    pub fn new(action: &Action, services: &HashMap<String, Service>) -> Result<Self, String> {
+        let service = services
+            .get(&action.service)
+            .ok_or(format!("Unknown service: {}", action.service))?;
+
+        match service.service_type {
+            ServiceType::RateLimit => Ok(Self::RateLimit(RateLimitAction::new(action, service)?)),
+            ServiceType::Auth => Ok(Self::Auth(AuthAction::new(action, service)?)),
+        }
+    }
+
+    pub fn service(&self) -> &str {
+        match self {
+            Self::Auth(auth_action) => auth_action.service(),
+            Self::RateLimit(rl_action) => rl_action.service(),
+        }
+    }
+
+    pub fn scope(&self) -> &str {
+        match self {
+            Self::Auth(auth_action) => auth_action.scope(),
+            Self::RateLimit(rl_action) => rl_action.scope(),
+        }
+    }
+
+    pub fn grpc_service(&self) -> Rc<GrpcService> {
+        match self {
+            Self::Auth(auth_action) => auth_action.get_grpcservice(),
+            Self::RateLimit(rl_action) => rl_action.get_grpcservice(),
+        }
+    }
+
+    pub fn merge(&mut self, other: &RuntimeAction) {
+        // only makes sense for rate limiting actions
+        if let Self::RateLimit(self_rl_action) = self {
+            if let Self::RateLimit(other_rl_action) = other {
+                self_rl_action.merge(other_rl_action)
+            }
+        }
+    }
```

Comment:
Well, not really. `RateLimitAction` can build descriptor entries and implements `merge` method. Also the `conditions_apply` is different because the predicates are now *guards* to allow descriptor entries to be included in the descriptor or not.